doc/src/manual/methods.md

```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
 "definition for Int"
 ```
 
+## Design Patterns with Parametric Methods
+
+
+While complex dispatch logic is not required for performance or usability,
+sometimes it can be the best way to express some algorithm.
+Here are a few common design patterns that come up sometimes when using dispatch in this way.
+
+### Extracting the type parameter from a super-type
+
+
+Here is the correct code template for returning the element-type `T`
+of any arbitrary subtype of `AbstractArray`:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{<:AbstractArray{T}}) where {T} = T
+```
+using so-called triangular dispatch.
+
+Another way, which used to be the only correct way before the advent of
+triangular dispatch, is:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{AbstractArray}) = Any
+eltype{T}(::Type{AbstractArray{T}}) = T
+eltype{T, N}(::Type{AbstractArray{T, N}}) = T
+eltype{A<:AbstractArray}(::Type{A}) = eltype(supertype(A))
+```
```

Comment:
This seems pretty complicated! More so than the old version, so we should either keep the old one or my update below, if it passes scrutiny.

I tried to update "my" version and this passes your tests:
```julia
f(::Type{<:AbstractArray{T}}) where {T} = T
f(::Type{<:AbstractArray}) = Any

@test f(Array{Integer, 3}) == Integer
@test f(Array{Integer}) == Integer
@test f(Array{Integer, N} where N<:Integer) == Integer
@test f(Array{T, 1} where T <: Integer) == Any
@test f(Array) == Any
@test f(Array{T} where T <: Integer) == Any
@test f(DenseArray{Integer}) == Integer
@test f(DenseArray{<:Integer}) == Any
```
The second methods seems to catch those `Any` cases.  Anything this misses?

On a more philosophical note, why not return `Integer` for `DenseArray{<:Integer}`? As all types which are subtypes of `DenseArray{<:Integer}` will have eltype `<:Integer`.